### PR TITLE
Production-ready content for MSAL Go

### PR DESCRIPTION
**Pull request by localden**

### CODE REVIEW

File: msal-go/index.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+---
+title: Microsoft Authentication Library (MSAL) for Go
+description: "An introduction to using Microsoft Authentication Library (MSAL) for Go."
+---
+
+# Microsoft Authentication Library (MSAL) for Go
+
+>[!NOTE]
+> icrosoft Authentication Library (MSAL) for Go is a new addition to the MSAL family of libraries. It has been made available in production-ready preview to gauge customer interest and to gather feedback from the community. We welcome all contributors (see [contribution guidelines in the library repository](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/CONTRIBUTING.md)) to help us improve the library.**
```
Comment (pmaytak) [inline msal-go/index.md]:
```suggestion
> Microsoft Authentication Library (MSAL) for Go is a new addition to the MSAL family of libraries. It has been made available in production-ready preview to gauge customer interest and to gather feedback from the community. We welcome all contributors (see [contribution guidelines in the library repository](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/dev/CONTRIBUTING.md)) to help us improve the library.
```

File: msal-go/index.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        // found a cached account, now see if an applicable token has been cached
+        result, err := publicClientApp.AcquireTokenSilent(context.Background(), []string{"your_scope"}, public.WithSilentAccount(userAccount))
+        accessToken := result.AccessToken
+    }
+    ```
+
+1. If there is no suitable token in the cache, or you choose to skip this step, now we can send a request to AAD to obtain a token.
+
+    ```go
+    result, err := publicClientApp.AcquireToken"ByOneofTheActualMethods"([]string{"your_scope"}, ...(other parameters depending on the function))
```
Comment (rayluo) [inline msal-go/index.md]: Many of the wording from this document sound very familiar to me. :-)

Half joking aside, this particular line is not currently syntactically valid. Was that intentional? Do you consider write it like below?
```suggestion
    result, err := publicClientApp.AcquireTokenByOneofTheActualMethods([]string{"your_scope"}, ...(other parameters depending on the function))
```

File: msal-go/design.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+We use this feature quite liberally to make clear what is using an internal package.  For example:
+
+```bash
+apps/internal/base - Only can be used by packages defined at apps/
+apps/internal/base/internal/storage - Only can be use by package client
+```
+
+## Public API
+
+The public API will be encapsulated in `apps/`.  `apps/` has 3 packages of interest to users:
```
Comment (bgavrilMS) [inline msal-go/design.md]:
```suggestion
The public API is encapsulated in `apps/`.  `apps/` has 3 packages of interest to users:
```

File: msal-go/design.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+- `public/` - This is what MSAL calls the Public Application Client (service client)
```
Comment (bgavrilMS) [inline msal-go/design.md]:
```suggestion
- `public/` - Covers the Public Application Client. These are apps that run on the user's computer, such as command line interface apps. 
```

File: msal-go/design.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+- `confidential/` - This is what MSAL calls the Confidential Application Client (service)
```
Comment (bgavrilMS) [inline msal-go/design.md]:
```suggestion
- `confidential/` - Covers the Confidential Application Client. These are apps that run on servers - web apps, web APIs which get tokens for users and web APIs which get tokens on behalf of themselves (service to service communication)

File: msal-go/design.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+- `cache/` - This provides the interfaces that must be implemented to create persistent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about `internal/`.
+
+### JSON Handling
+
+JSON must be handled specially in our app. The basics are, if we receive fields that our structs do not contain, we cannot drop them.  We must send them back to the service.
```
Comment (bgavrilMS) [inline msal-go/design.md]:
```suggestion
JSON must be handled specially in our app. The basics are, if we receive fields that our structs do not contain, we cannot drop them, to ensure backwards and future compatibility.
```

File: msal-go/design.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to `ops.REST`
+- Add the higher level manipulations to `oauth.Token`
+- Add your logic to the `app/\<client\>` and access the services via your `oauth.Token`
+
+## Notable Differences To Other Clients
+
+### TBD: Confidential applications needs to handle multiple users without one big cache
```
Comment (bgavrilMS) [inline msal-go/design.md]:
```suggestion
### Confidential applications needs to handle multiple users without one big cache
```

File: msal-go/index.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+You can view the [dev apps](https://github.com/AzureAD/microsoft-authentication-library-for-go/tree/dev/apps/tests/devapps) on how to use MSAL Go with various application types in various scenarios. For more detailed information, please refer to the [wiki](https://github.com/AzureAD/microsoft-authentication-library-for-go/wiki).
+
+## Releases
+
+For a full list of library releases, refer to the [Releases](https://github.com/AzureAD/microsoft-authentication-library-for-go/releases) section in the library source code repository.
+
+## Community help and support
+
+We use [Stack Overflow](https://stackoverflow.com/questions/tagged/azure-ad-msal) to work with the community on supporting Azure Active Directory and its SDKs, including this one! We highly recommend you ask your questions on Stack Overflow. You can also browse existing questions to see if someone has encountered the problem before. Please use the `azure-ad-msal` tag when asking your questions.
```
Comment (bgavrilMS) [inline msal-go/index.md:92]: Do we really monitor Stack Overflow?

**Review by bgavrilMS — APPROVED** [review]

File: msal-go/index.md (hunk 8 above)
Comment (rayluo) [reply msal-go/index.md:92]: I believe there are some support engineers from Microsoft monitoring StackOverflow. Not sure who and how often. MSAL team here historically does not monitor StackOverflow. That being said, occasionally when I receive 1:1 questions on Teams, and if the question is generic, I suggest the question asker to post their question on StackOverflow, and paste me the link if it got unanswered after a day or two.

TL;DR. I would still suggest keep this StackOverflow paragraph.

Comment (localden) [reply msal-go/index.md:92]: I am monitoring Stack Overflow @bgavrilMS @rayluo 😄

File: msal-go/index.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    if len(accounts) > 0 {
+        // Assuming the user wanted the first account
+        userAccount = accounts[0]
+        // found a cached account, now see if an applicable token has been cached
+        result, err := publicClientApp.AcquireTokenSilent(context.Background(), []string{"your_scope"}, public.WithSilentAccount(userAccount))
+        accessToken := result.AccessToken
+    }
+    ```
+
+1. If there is no suitable token in the cache, or you choose to skip this step, now we can send a request to AAD to obtain a token.
```
Comment (Dickson-Mwendia) [inline msal-go/index.md]: If there's no suitable token in the cache, or you choose to skip this step, send a request to Azure AD to obtain a token. There are different methods to acquire a token based on your application type and scenario. Here, we demonstrate a placeholder flow.

File: msal-go/index.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    ```go
+    result, err := publicClientApp.AcquireTokenByOneofTheActualMethods([]string{"your_scope"}, ...(other parameters depending on the function))
+    if err != nil {
+        log.Fatal(err)
+    }
+    accessToken := result.AccessToken
+    ```
+
+You can view the [dev apps](https://github.com/AzureAD/microsoft-authentication-library-for-go/tree/dev/apps/tests/devapps) on how to use MSAL Go with various application types in various scenarios. For more detailed information, please refer to the [wiki](https://github.com/AzureAD/microsoft-authentication-library-for-go/wiki).
```
Comment (Dickson-Mwendia) [inline msal-go/index.md]: You can view the [developer sample apps](https://github.com/AzureAD/microsoft-authentication-library-for-go/tree/dev/apps/tests/devapps) on how to use MSAL Go with various application types in different scenarios.

Comment (Dickson-Mwendia) [reply msal-go/index.md]: I'd recommend doing away with this pointer to the wiki

File: msal-go/index.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+If you find a bug or have a feature request, please open a new issue in the [Issues](https://github.com/AzureAD/microsoft-authentication-library-for-go/issues) section.
+
+## Submit feedback
+
+If you have any library feedback, make sure to submit your feature requests and bug reports [on GitHub](https://github.com/AzureAD/microsoft-authentication-library-for-go/issues).
+
+## Security library
+
+This library controls how users sign-in and access services. We recommend you always take the latest version of our library in your app when possible. We use [semantic versioning](http://semver.org) so you can control the risk associated with updating your app. As an example, always downloading the latest minor version number (e.g. x.*y*.x) ensures you get the latest security and feature enhancements but our API surface remains the same. You can always see the latest version and release notes under the Releases tab of GitHub.
```
Comment (Dickson-Mwendia) [inline msal-go/index.md]: This library controls how users sign-in and access services. We recommend using the latest version of our library in your app when possible. We use [semantic versioning](http://semver.org) so you can control the risk associated with updating your app. As an example, always downloading the latest minor version number (e.g. x.*y*.x) ensures you get the latest security and feature enhancements but our API surface remains the same. You can always see the latest version and release notes under the Releases tab of GitHub.

File: msal-go/TOC.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+- name: Microsoft Authentication Library for Go
+  href: index.md
+- name: Design Guide
```
Comment (Dickson-Mwendia) [inline msal-go/TOC.yml]: - name: Design guide

File: msal-go/TOC.yml [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  href: design.md
+- name: Error Design
```
Comment (Dickson-Mwendia) [inline msal-go/TOC.yml]: - name: Error design

File: msal-go/design.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
+---
+title: Microsoft Authentication Library for Go Design Guide
+description: "Design guidelines for the Microsoft Authentication Library for Go."
+---
+
+# Microsoft Authentication Library for Go Design Guide
+
+## General Structure
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: ## General structure

File: msal-go/design.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+Public Surface
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: Public surface

File: msal-go/design.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+- `public/` - Covers the Public Application Client. These are apps that run on the user's computer, such as command line interface apps. 
+- `confidential/` - Covers the Confidential Application Client. These are apps that run on servers - web apps, web APIs which get tokens for users and web APIs which get tokens on behalf of themselves (service to service communication)
+- `cache/` - This provides the interfaces that must be implemented to create persistent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about `internal/`.
+
+### JSON Handling
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: ### JSON handling

File: msal-go/design.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+To handle that, we use our own custom `json` package that handles this.
+
+### Backend communication
+
+Communication to the backends is done via the `requests/` package. `oauth.Token` is the client for all communication.
+
+`oauth.Token` communicates via REST calls that are encapsulated in the `ops/` client.
+
+## Adding A Feature
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: ## Adding a feature

File: msal-go/design.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to `ops.REST`
+- Add the higher level manipulations to `oauth.Token`
+- Add your logic to the `app/\<client\>` and access the services via your `oauth.Token`
+
+## Notable Differences To Other Clients
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: ## Notable differences to other clients

File: msal-go/design.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+The public API is encapsulated in `apps/`.  `apps/` has 3 packages of interest to users:
+
+- `public/` - Covers the Public Application Client. These are apps that run on the user's computer, such as command line interface apps. 
+- `confidential/` - Covers the Confidential Application Client. These are apps that run on servers - web apps, web APIs which get tokens for users and web APIs which get tokens on behalf of themselves (service to service communication)
+- `cache/` - This provides the interfaces that must be implemented to create persistent caches for any MSAL client
+
+## Internals
+
+In this section we will be talking about `internal/`.
```
Comment (Dickson-Mwendia) [inline msal-go/design.md]: In this section, we will be talking about `internal/`.

File: msal-go/design.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+This is the general way to add a new feature to MSAL:
+
+- Add the REST calls to `ops.REST`
+- Add the higher level manipulations to `oauth.Token`
+- Add your logic to the `app/\<client\>` and access the services via your `oauth.Token`
+
+## Notable Differences To Other Clients
+
+### Confidential applications needs to handle multiple users without one big cache
```
Comment (Dickson-Mwendia) [inline msal-go/design.md:77]: ### Confidential applications need to handle multiple users without one big cache

File: msal-go/error-design.md [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+- Detect errors that are transitory and can be retried
+- Allow the user to identify certain errors that the program can respond to, such a informing the user for the need to do an enrollment
+  
+## Implementing client-side errors
+
+Client side errors indicate a misconfiguration or passing of bad arguments that is non-recoverable. Retrying isn't possible.
+
+These errors can simply be standard Go errors created by `errors.New()` or `fmt.Errorf()`. If down the line we need a custom error, we can introduce it, but for now the error messages just need to be clear on what the issue was.
+
+## Implementing Service Side Errors
```
Comment (Dickson-Mwendia) [inline msal-go/error-design.md]: ## Implementing service-side errors

File: msal-go/error-design.md [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+```go
+return nil, errors.CallErr{
+    Req:  req,
+    Resp: reply,
+    Err:  fmt.Errorf("http call(%s)(%s) error: reply status code was %d:\n%s", req.URL.String(), req.Method, reply.StatusCode, ErrorResponse), //ErrorResponse is the json body extracted from the http response
+    }
+```
+
+## Future Decisions
```
Comment (Dickson-Mwendia) [inline msal-go/error-design.md]: ## Future decisions

File: msal-go/error-design.md [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+```go
+type MyCustomErr struct {
+  Msg string
+}
+func (m MyCustomErr) Error() string { // This implements "error"
+  return m.Msg
+}
+```
+
+### MSAL error goals
```
Comment (Dickson-Mwendia) [inline msal-go/error-design.md]: Thoughts on placing these points at the top of the article, - just below the sentence in ##background?

**Review by Dickson-Mwendia — CHANGES_REQUESTED** [review]
@localden a few minor suggestions here. In line with docs' contributor guide all H1, H2s/H3s should be sentence case.  Otherwise, it looks good. I'll make a few content updates once this is merged. Thanks!
